cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cassert>
+#include <limits>
+#include <functional>
+#include <boost/serialization/strong_typedef.hpp>
+
+namespace cudf {
+namespace fp {
+
+BOOST_STRONG_TYPEDEF(int32_t, scale_type)
+
+enum Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type{
+        return scale_type{-1 * scale};
+    }
+
+    // perform this operation when constructing with - scale (after negating scale)
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) {
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
```

Comment:
This question applies to the other operators as well.

Reply:
The compiler will automatically do this optimization for `Radix` since it's compile time info.

It's not worth it for the runtime scale info. For the same reason we don't have conditional checks for adding with zero or multiplying with one. All those runtime checks will just make everything slower on average.